dbgd.c

```diff
@@ -178,132 +175,225 @@ void dbgd_init(void)
+    typedef struct {
+      uint32_t address;
+    } ResponseData;
 
-    return DBG__RESPONSE__TYPE__OK;
+    ResponseData response;
+    response.address = (uint32_t)malloc(request.size);
+
+    send(fd, &response, sizeof(response), 0);
+
+    return 0;
 }
 
-static int dbgd_free(Dbg__Request *req, Dbg__Response *res)
+static int dbgd_free(int fd)
 {
-    if (!req->has_address)
-        return DBG__RESPONSE__TYPE__ERROR_INCOMPLETE_REQUEST;
+    bool success;
+
+    typedef struct {
+      uint32_t address;
+    } RequestData;
 
-    free((void*)req->address);
+    RequestData request;
+    success = ensure_recv(fd, &request, sizeof(request));
+    assert(success);
 
-    return DBG__RESPONSE__TYPE__OK;
+    free((void*)request.address);
+
+    return 0;
 }
 
-static int dbgd_mem_read(Dbg__Request *req, Dbg__Response *res)
+static int dbgd_mem_read(int fd)
 {
-    if (!req->has_address || !req->has_size)
-        return DBG__RESPONSE__TYPE__ERROR_INCOMPLETE_REQUEST;
+    bool success;
+
+    typedef struct {
+      uint32_t address;
+      uint32_t size;
+    } RequestData;
 
-    res->address = req->address;
-    res->has_address = 1;
+    typedef struct {
+      uint8_t data[1];
+    } ResponseData;
 
-    res->data.len  = req->size;
-    res->data.data = get_transfer_buffer(res->data.len);
-    res->has_data = 1;
+    RequestData request;
+    success = ensure_recv(fd, &request, sizeof(request));
+
+    size_t response_data_length = sizeof(ResponseData) - 1 + request.size;
+    ResponseData* response_data = malloc(response_data_length);
 
     unsigned int i = 0;
-    unsigned int s = req->size;
+    unsigned int s = request.size;
 
     while(s >= 4) {
-      *(uint32_t*)&res->data.data[i] = *(volatile uint32_t*)(req->address + i);
+      *(uint32_t*)&response_data->data[i] = *(volatile uint32_t*)(request.address + i);
       i += 4;
       s -= 4;
     }
     while(s >= 2) {
-      *(uint16_t*)&res->data.data[i] = *(volatile uint16_t*)(req->address + i);
+      *(uint16_t*)&response_data->data[i] = *(volatile uint16_t*)(request.address + i);
       i += 2;
       s -= 2;
     }
     while(s >= 1) {
-      *(uint8_t*)&res->data.data[i] = *(volatile uint8_t*)(req->address + i);
+      *(uint8_t*)&response_data->data[i] = *(volatile uint8_t*)(request.address + i);
       i += 1;
       s -= 1;
     }
 
-    return DBG__RESPONSE__TYPE__OK;
+    debugPrint("Responding %u / %u / %u bytes\n", sizeof(ResponseData), response_data_length, request.size);
+    send(fd, response_data, response_data_length, 0);
+
+    free(response_data);
+
+    return 0;
 }
 
-static int dbgd_mem_write(Dbg__Request *req, Dbg__Response *res)
+static int dbgd_mem_write(int fd)
 {
-    if (!req->has_address || !req->has_data)
-        return DBG__RESPONSE__TYPE__ERROR_INCOMPLETE_REQUEST;
+    bool success;
+
+    typedef struct {
+      uint32_t address;
+      uint32_t data_length;
+    } RequestHeader;
+
+    typedef struct {
+      uint8_t data[1];
+    } RequestData;
```

Comment:
This style of request definition is a bit non-extensible.

What if we introduce commands which require more back-and-forth or multiple arguments of variable size?

Maybe `RequestHeader` and `ResponseHeader` should be enough, and the data part is always parsed manually?